Update fsmonitor watchman scripts

kewillford opened the pull request:
When running Git commands quickly -- such as in a shell script or the
test suite -- the Git commands frequently complete and start again
during the same second. The example fsmonitor hooks to integrate with
Watchman truncate the nanosecond times to seconds. In principle, this is
fine, as [Watchman claims to use inclusive comparisons](https://github.com/facebook/watchman/blob/master/query/since.cpp#L35-L39). The result
should only be an over-representation of the changed paths since the
last Git command.

However, Watchman's own documentation claims ["Using a timestamp is prone
to race conditions in understanding the complete state of the file tree"](https://facebook.github.io/watchman/docs/clockspec.html).
All of their documented examples use a "clockspec" that looks like
'c:123:234'. Git should eventually learn how to store this type of
string to provide a stronger integration, but that will be a more
invasive change.

When using GIT_TEST_FSMONITOR="$(pwd)/t7519/fsmonitor-watchman", scripts
such as t7519-wtstatus.sh fail due to these race conditions. In fact,
running any test script with GIT_TEST_FSMONITOR pointing at
t/t7519/fsmonitor-wathcman will cause failures in the test_commit
function. The 'git add "$indir$file"' command fails due to not enough
time between the creation of '$file' and the 'git add' command.

For now, subtract one second from the timestamp we pass to Watchman.
This will make our window large enough to avoid these race conditions.
Increasing the window causes tests like t7519-wtstatus.sh to pass.

When the integration was introduced in def4376 (fsmonitor: add a
sample integration script for Watchman, 2018-09-22), the query included
an expression that would ignore files created and deleted in that
window. The performance reason for this change was to ignore temporary
files created by a build between Git commands. However, this causes
failures in script scenarios where Git is creating or deleting files
quickly.

When using GIT_TEST_FSMONITOR as before, t2203-add-intent.sh fails
due to this add-and-delete race condition.

By removing the "expression" from the Watchman query, we remove this
race condition. It will lead to some performance degradation in the case
of users creating and deleting temporary files inside their working
directory between Git commands. However, that is a cost we need to pay
to be correct.

derrickstolee commented on fsmonitor.c:
```diff
@@ -203,7 +207,7 @@ void refresh_fsmonitor(struct index_state *istate)
 
 	/* Now that we've updated istate, save the last_update time */
-	istate->fsmonitor_last_update = last_update;
+	istate->fsmonitor_last_update = last_update - 1000000000;
```
Please comment on the change here and why we are subtracting.

I changed the example in `templates/hooks--fsmonitor-watchman.sample` and `t/t7519/fsmonitor-watchman` instead, as they seem to be watchman-specific adjustments instead of something universal to all `fsmonitor` integrations.

derrickstolee replied:
Also, `t/t7519/fsmonitor-watchman` has the following query:

```
["query", "$git_work_tree", {
			"since": $time,
			"fields": ["name"],
			"expression": ["not", ["allof", ["since", $time, "cclock"], ["not", "exists"]]]
		}]
```

but the template doesn't have the `expression` line. That seems to be the thing causing the add/delete nullifications, right?

kewillford replied:
Yes that expression is what is filtering out the changes since X that no longer exist.

kewillford replied:
I'll move some things and clean this up and put it on it's own PR so we can test without my other changes.

derrickstolee replied:
(Sorry, I think I had the files backwards as to which has the expression and which does not.)

Commit:
fsmonitor: fix watchman integration

When running Git commands quickly -- such as in a shell script or the
test suite -- the Git commands frequently complete and start again
during the same second. The example fsmonitor hooks to integrate with
Watchman truncate the nanosecond times to seconds. In principle, this is
fine, as Watchman claims to use inclusive comparisons [1]. The result
should only be an over-representation of the changed paths since the
last Git command.

However, Watchman's own documentation claims "Using a timestamp is prone
to race conditions in understanding the complete state of the file tree"
[2]. All of their documented examples use a "clockspec" that looks like
'c:123:234'. Git should eventually learn how to store this type of
string to provide a stronger integration, but that will be a more
invasive change.

When using GIT_TEST_FSMONITOR="$(pwd)/t7519/fsmonitor-watchman", scripts
such as t7519-wtstatus.sh fail due to these race conditions. In fact,
running any test script with GIT_TEST_FSMONITOR pointing at
t/t7519/fsmonitor-wathcman will cause failures in the test_commit
function. The 'git add "$indir$file"' command fails due to not enough
time between the creation of '$file' and the 'git add' command.

For now, subtract one second from the timestamp we pass to Watchman.
This will make our window large enough to avoid these race conditions.
Increasing the window causes tests like t7519-wtstatus.sh to pass.

When the integration was introduced in def437671 (fsmonitor: add a
sample integration script for Watchman, 2018-09-22), the query included
an expression that would ignore files created and deleted in that
window. The performance reason for this change was to ignore temporary
files created by a build between Git commands. However, this causes
failures in script scenarios where Git is creating or deleting files
quickly.

When using GIT_TEST_FSMONITOR as before, t2203-add-intent.sh fails
due to this add-and-delete race condition.

By removing the "expression" from the Watchman query, we remove this
race condition. It will lead to some performance degradation in the case
of users creating and deleting temporary files inside their working
directory between Git commands. However, that is a cost we need to pay
to be correct.

[1] https://github.com/facebook/watchman/blob/master/query/since.cpp#L35-L39
[2] https://facebook.github.io/watchman/docs/clockspec.html

Helped-by: Derrick Stolee <dstolee@microsoft.com>
Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

wilbaker commented:
> This also means that the filter for adds that are then removed needs to be deleted because the add might have happened before the previous query and fsmonitor still needs to know about the remove that happens.

Could you provide a few more details on this scenario?  Does "adds that are then removed" mean new files that are quickly created and deleted, or is it something to do with `git add`?

wilbaker commented on t/t7519/fsmonitor-watchman:
```diff
@@ -64,8 +65,7 @@ sub launch_watchman {
 			"since": $time,
-			"fields": ["name"],
-			"expression": ["not", ["allof", ["since", $time, "cclock"], ["not", "exists"]]]
+			"fields": ["name"]
```
Can you update the above comment? This part doesn't appear to be true anymore:

`# To accomplish this, we're using the "since" generator to use the`

wilbaker replied:
> Could you provide a few more details on this scenario? Does "adds that are then removed" mean new files that are quickly created and deleted, or is it something to do with git add?

I think I just realized you were referring to this code, in that case, please ignore my earlier comment.

wilbaker replied:
Also, what is the downside of getting the names of files that were transient.  Is it just a perf issue?

kewillford replied:
>what is the downside of getting the names of files that were transient. Is it just a perf issue?

Yes it would just mean that more file names would be returned that would need to be processed and in most cases these kind of files are ones that don't need to be because they were not in the index because they were new files (untracked) but now don't exist so won't be in the git status.

wilbaker commented on templates/hooks--fsmonitor-watchman.sample:
```diff
@@ -63,8 +64,7 @@ sub launch_watchman {
 			"since": $time,
-			"fields": ["name"],
-			"expression": ["not", ["allof", ["since", $time, "cclock"], ["not", "exists"]]]
+			"fields": ["name"]
```
Same comment: Update the above comment to better match the query we're performing here.

wilbaker commented on templates/hooks--fsmonitor-watchman.sample:
```diff
@@ -22,7 +22,8 @@ my ($version, $time) = @ARGV;
 if ($version == 1) {
 	# convert nanoseconds to seconds
-	$time = int $time / 1000000000;
+	# substract one second to make sure watchman will return all changes
```
Do you know if the "one second" issue is Mac specific? (that would matter more for upstreaming than this PR)

wilbaker replied:
Related you mentioned:

> watchman returns are not inclusive of the time that is passed to it

How did you track this down?  If you found this in the Watchman source could you provide a link to it?

jrbriggs replied:
(not the author, but answering because I've been following along and want to double-check my understanding)

We need to back off a second because we're snapping these queries to the second boundary, and the `since` expression in watchman performs a greater-than check (https://facebook.github.io/watchman/docs/expr/since.html). If we don't back off one unit of what we're querying by, and in this case it is seconds, x>x is always false.

@kewillford @derrickstolee did I get that correct?

kewillford replied:
Yes and the documentation on [clockspec](https://facebook.github.io/watchman/docs/clockspec.html) says "Using a timestamp is prone to race conditions in understanding the complete state of the file tree."

jrbriggs replied:
Is it possible to use the abstract clockid to help guard against these race conditions?

derrickstolee replied:
@jrbriggs: @kewillford and I talked about using the clock time string that Watchman provides, but it requires adding an extension to the index. Definitely something to look into, but not appropriate for this fix.

derrickstolee commented:
@kewillford I was looking into this, too, and created a commit (c5f96de1) that has very similar changes to yours. Perhaps you could just take the commit message instead? This seems important to upstream.

```
fsmonitor: fix watchman integration
    
    When running Git commands quickly -- such as in a shell script or our
    test suite -- the Git commands frequently complete and start again
    during the same second. Our example fsmonitor hooks to integrate with
    Watchman truncate the nanosecond times to seconds. In principle, this is
    fine, as Watchman claims to use inclusive comparisons [1]. The result
    should only be an over-representation of the changed paths since the
    last Git command.
    
    However, Watchman's own documentation claims "Using a timestamp is prone
    to race conditions in understanding the complete state of the file tree"
    [2]. All of their documented examples use a "clockspec" that looks like
    'c:123:234'. Git should eventually learn how to store this type of
    string to provide a stronger integration, but that will be a more
    invasive change.
    
    When using GIT_TEST_FSMONITOR="$(pwd)/t7519/fsmonitor-watchman", scripts
    such as t7519-wtstatus.sh fail due to these race conditions. In fact,
    running any test script with GIT_TEST_FSMONITOR pointing at
    t/t7519/fsmonitor-wathcman will cause failures in the test_commit
    function. The 'git add "$indir$file"' command fails due to not enough
    time between the creation of '$file' and the 'git add' command.
    
    For now, subtract one second from the timestamp we pass to Watchman.
    This will make our window large enough to avoid these race conditions.
    Increasing the window causes tests like t7519-wtstatus.sh to pass.
    
    When the integration was introduced in def437671 (fsmonitor: add a
    sample integration script for Watchman, 2018-09-22), the query included
    an expression that would ignore files created and deleted in that
    window. The performance reason for this change was to ignore temporary
    files created by a build between Git commands. However, this causes
    failures in script scenarios where Git is creating or deleting files
    quickly.
    
    When using GIT_TEST_FSMONITOR as before, t2203-add-intent.sh fails
    due to this add-and-delete race condition.
    
    By removing the "expression" from the Watchman query, we remove this
    race condition. It will lead to some performance degradation in the case
    of users creating and deleting temporary files inside their working
    directory between Git commands. However, that is a cost we need to pay
    to be correct.
    
    [1] https://github.com/facebook/watchman/blob/master/query/since.cpp#L35-L39
    [2] https://facebook.github.io/watchman/docs/clockspec.html
```

derrickstolee commented:
/azp run Microsoft.git
